### ISSUE TITLE
Making additional client certifications?

### ISSUE THREAD

**Issue by mjosh20** [google/simian#11]
I was able to add a few client to Simian w/o issue.  Now I'm looking to do a bigger deployment, but how can I make more certificates w/o have to delete all the certs from the app engine, redoing all the private keys, and have to re-do the clients that are currently working.  I tried doing the export CA="my password" and then running the command to make the certs, but I got an error about not finding the key.  Any help would be greatly appreciated...Thanks again for all your work on Simian!!!!

**Comment by juofmgoblue** [comment]
TL;DR:  supporting multiple certificate authorities in Simian is not straightforward, and not documented, and can only be done if you have significant experience with Python and some App Engine development chops.  


We have functionality to support multiple certificate authorities, and we've used it internally to migrate from one CA to another, but we've never tested it in our open source app, and thus never documented it.  It may or may not work.  Unfortunately we cannot support it at this time, as it would take a lot of effort to streamline for non-developers.  But I'll provide pointers in case you have the requisite experience to proceed.

GetCaId() loads CA_ID from the settings/config file, and defaults to None if that (undocumented) variable doesn't exist: https://github.com/google/simian/blob/master/src/simian/auth/util.py#L65

The client will use this value (if present) to instruct the server which CA it's using, which effects which private_key/public_cert pair is loaded from App Engine Datastore's "Settings" model.  You should be able to use Datastore Viewer to see the server_private_key_pem and other keys in that model today, as uploaded in the admin web UI.  You'd have to write some Python code to use the Datastore API to write additional entities to the settings table, similar to the existing ones but with a prefix matching CA_ID in the clients using the new CA: https://github.com/google/simian/blob/master/src/simian/auth/util.py#L124

If that all makes sense and you need a little more guidance, I'm happy to answer questions.  If it seems too difficult, it's because it is, and it would take many, many hours to make easier, unfortunately.  We can leave this issue open to use as a feature request for making this process easier.

**Comment by mjosh20** [comment]
Sorry, thats not what I was trying to ask. My fault.   

I want to be able to make my own certificates in an on-demand manner.  I followed the wiki about openssl and I made certificates for all my test machines and everything worked.  Now I want to deploy to more clients, but I can't seem to get the process to run w/o starting over with the whole how-to.  

That would require me to delete all my current clients and the keys from the data store and start re-adding the clients that are working.  

This is the part I need to be able to do w/o remaking the CA identity, server identity, etc:

for serial in $(cat serials.txt) ; do 

  rm -f client.csr

  openssl req -config openssl.cnf -sha1 -new -nodes \
  -out client.csr \
  -keyout private/${serial}.pem \
  -subj "/CN=${serial}"

  openssl ca -config openssl.cnf -in client.csr \
  -md sha1 -days 3650 -notext \
  -cert newcerts/ca_public_cert.pem -keyfile private/ca.key \
  -out newcerts/${serial}.pem \
  -policy policy_anything \
  -passin env:CAPASS -batch

done


Does that make more sense?

**Comment by johnrandolph** [comment]
Hi mjosh20

can you verify that all of the files like private/ca.key openssl.cnf etc still exist and that you are in the right portion of the directory tree? cd to the top of the path that contains all of those subdirectories like newcerts/ private/ etc and verify what you see in there. perhaps paste it here.

also prior to running the openssl commands, are you sure you entered:

export CAPASS="your passsword"

and not just

CAPASS="your password" 

?

the first one (export...) is what you need.

**Comment by mjosh20** [comment]
....................++++++
writing new private key to 'private/00000000001.pem'
-----
Using configuration from openssl.cnf
unable to load CA private key
59460:error:0906A065:PEM routines:PEM_do_header:bad decrypt:/SourceCache/OpenSSL098/OpenSSL098-52.8.1/src/crypto/pem/pem_lib.c:428:
Generating a 1024 bit RSA private key
..++++++
..........++++++

I think it has something to do w/ the fact that I don't remake the server identity in the process.  I go from the export CAPASS to running the script to make the client certs.  

Thoughts?

**Comment by johnrandolph** [comment]
you need to verify that newcerts/ca_public_cert.pem and private/ca.key still exist.

you should perform these new client cert generations from the top level directory in which those newcerts/ and private/ directories descend from.

**Comment by mjosh20** [comment]
They still exist in the correct folders.
![ca key and ca public cert](https://cloud.githubusercontent.com/assets/10607571/6066744/81119be2-ad3d-11e4-8ba7-ed27bfc71dd1.png)

I still get the error no matter what I try.

**Comment by mjosh20** [comment]
I deleted CA and all the certificates from the appengine and started fresh.  I made ~10 client certs w/o any issue.  I then try to make a few additional ones.  I run the export CAPASS='MY PASSWORD HERE', then I run the the script that generates the certs.  I no longer get the error I was getting but I get the following:

            X509v3 Basic Constraints: 
                CA:FALSE
            Netscape Comment: 
                OpenSSL Generated Certificate

The private key appears to have generated but the other one is blank.  I assume its due to the CA: False part in the output.  I'm positive its still in the CA/private.  Is there a way to make a 2nd round of keys w/o remaking the CA.key?

**Comment by johnrandolph** [comment]
Hi mjosh20
there should be no problem running the client private key and create cert with client private key + signed w/ca private key commands in more than one iteration.

The paste you have supplied is fine, CA:FALSE, if it is from openssl x509 -in clientcert.pem -text . The client certs should not have CA:TRUE because they are not CA certs and should not be used for signing. 

The CA cert will show CA:TRUE, example:
:/tmp/CA$ openssl x509 -in newcerts/ca_public_cert.pem  -text 
...
```
            X509v3 Basic Constraints: 
                CA:TRUE
```
After your initial pass of 10 client certs did you populate serials.txt with more serial numbers, assuming that you are using our wiki cli cut/paste directly?

**Comment by mjosh20** [comment]
Yes, I added more of them to the end of it.  Your assumption is correct.

Additional note: The one client cert is coming out completely blank.  And the other one is similar to the first run but not exactly the same.

**Comment by johnrandolph** [comment]
Ok, so by your wording "I added more of them to the end of it" it sounds like you are probably instructing openssl to re-create private keys and signed certs for each client, including the ones you already generated this for. You don't want to do this.

Part of this confusion may be because we suggested that you could name the subject of your host after the Apple serial number of the hardware and refer to serials.txt etc. However each x509 cert also has a serial number field which is a x509 cert property and has no relation to the hardware. 

So if you start with this list
serial1
serial2
serial3

And generate keys/certs, but then append

serial4
serial5

You will re-create serial1,2,3 again (or possibly it will produce errors) and also generate new serial4,5.

The way that shell loop is written is such that you really only want serials.txt to contain the new items to work on, not all the prior items too. We could make our shell snippet smarter about repeats, but then you're starting to get more complicated...

can you try just making serials.txt contain new items to create and see if it works?

**Comment by johnrandolph** [comment]
I revised our wiki documentation to try to make it more clear that by throwing names into a text file and iterating over them with shell we are choosing commonName values for each certificate, and to remove any confusion that it's related to the x509 serial number.

**Comment by mjosh20** [comment]
That worked.  I noticed the changes when I copied them and it said the file didn't exist, then I got your message. :)

The new code will work just like before, except I need to change the name of my serials.txt correct?

**Comment by johnrandolph** [comment]
Yes, all I did was go into the text and change serials.txt to cns.txt
anywhere I saw it, and also renamed the variable serial to cn. I also added
some extra wording to clarify that while we are suggesting that perhaps you
could name each certificate with the CN=machineserialnumber, this is only a
suggestion and is not related to the x509 certificate serial number, which
is maintained by openssl. You seed it at "1" with the "echo 000001 >
serial" line earlier in the instructions.
